src/Bicep.Core/Syntax/SyntaxTreeFactory.cs

```diff
@@ -0,0 +1,85 @@
+using Bicep.Core.Extensions;
+using Bicep.Core.FileSystem;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax.Converters;
+using Bicep.Core.Text;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Bicep.Core.Syntax
+{
+    public static class SyntaxTreeFactory
+    {
+        public static SyntaxTree CreateSyntaxTree(Uri fileUri, string fileContents, bool isEntryFile) => isEntryFile
+            ? CreateBicepSyntaxTree(fileUri, fileContents)
+            : CreateSyntaxTree(fileUri, fileContents);
+
+        public static SyntaxTree CreateSyntaxTree(Uri fileUri, string fileContents) =>
+            PathHelper.HasBicepExtension(fileUri) || !PathHelper.HasAnyExtension(fileUri)
+                ? CreateBicepSyntaxTree(fileUri, fileContents)
+                : CreateJsonSyntaxTree(fileUri, fileContents);
+
+        public static SyntaxTree CreateBicepSyntaxTree(Uri fileUri, string fileContents)
+        {
+            var parser = new Parser(fileContents);
+            var lineStarts = TextCoordinateConverter.GetLineStarts(fileContents);
+            
+            return new SyntaxTree(fileUri, lineStarts, parser.Program());
+        }
+
+        public static JsonSyntaxTree CreateJsonSyntaxTree(Uri fileUri, string fileContents)
+        {
+            var lineStarts = TextCoordinateConverter.GetLineStarts(fileContents);
+
+            try
+            {
+                var templateObject = JObject.Parse(fileContents, new JsonLoadSettings
+                {
+                    CommentHandling = CommentHandling.Ignore,
+                    LineInfoHandling = LineInfoHandling.Load,
+                });
+
+                var programSyntax = ProgramSyntaxConverter.ConvertToProgramSyntaxWithTypeInfoOnly(templateObject);
+
+                if (!Template.TryFromJson<Template>(fileContents, out var _))
+                {
+                    return CreateErrorJsonSyntaxTree(fileUri, lineStarts, "Failed to desirialize the template.", templateObject);
+                }
+
+                return new JsonSyntaxTree(fileUri, lineStarts, templateObject, programSyntax);
+            }
+            catch (JsonReaderException e)
+            {
+                return CreateErrorJsonSyntaxTree(fileUri, lineStarts, e.Message, e.LineNumber, e.LinePosition);
+            }
+            catch (SyntaxConversionException e)
+            {
+                return CreateErrorJsonSyntaxTree(fileUri, lineStarts, e.Message, e.LineNumber, e.LinePosition);
+            }
+        }
+
+        private static JsonSyntaxTree CreateErrorJsonSyntaxTree(Uri fileUri, ImmutableArray<int> lineStarts, string errorMessage, int errorLineNumber, int errorLinePosition)
+        {
+            var offset = TextCoordinateConverter.GetOffset(lineStarts, errorLineNumber - 1, errorLinePosition - 1);
+            var diagnostic = DiagnosticBuilder.ForPosition(new TextSpan(offset, 0)).InvalidJsonTemplate(errorMessage);
+            var program = new ProgramSyntax(
```

Comment:
>ProgramSyntax [](http://example.com/codeflow?start=30&length=13)

Do we need to create a busted ProgramSyntax in this case at all? Existing module logic has a pattern for producing errors when file doesn't exist or something else is wrong with reading it.

Reply:
Yes, it is needed here as it covers a different scenario where there's no problem reading the template file, but the file contents are invalid. The busted `ProgramSyntax` is used as a diagnostic container from which the language server can read in order to publish diagnostics for the template file.

Reply:
Would we still need it if we didn't bother providing detailed errors about the JSON template? (This is related to my other comment.)

Reply:
This is removed in the new implementation.